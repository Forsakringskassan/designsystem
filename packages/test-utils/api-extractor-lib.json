{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
