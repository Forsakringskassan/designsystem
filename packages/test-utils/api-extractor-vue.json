{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
