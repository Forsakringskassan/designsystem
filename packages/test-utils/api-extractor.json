{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
