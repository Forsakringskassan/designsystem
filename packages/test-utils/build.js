const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
