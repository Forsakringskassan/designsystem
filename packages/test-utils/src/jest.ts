/* imported for side-effects only */
import "./matchers";

export * from "./lib";
