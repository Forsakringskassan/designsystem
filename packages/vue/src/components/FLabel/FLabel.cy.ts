import { type DefineComponent, defineComponent } from "vue";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import { FLabelPageObject } from "../../cypress";
import FLabel from "./FLabel.vue";

function createComponent(template: string): DefineComponent {
    return defineComponent({
        template,
        components: {
            FLabel,
        },
    });
}

const defaultTemplate = /* HTML */ `
    <f-label id="label" for="input-field">
        Etikett
        <template #description="{ descriptionClass, formatDescriptionClass }">
            <span :class="descriptionClass"> Hjälptext </span>
            <span :class="formatDescriptionClass"> Formatbeskrivning </span>
        </template>
        <template #error-message> Felmeddelande </template>
    </f-label>
    <input id="input-field" type="text" class="text-field__input" />
`;

describe("FLabel", () => {
    beforeEach(() => {
        cy.mount(createComponent(defaultTemplate));
    });

    it("should provide a page object that can access any necessary elements", () => {
        const label = new FLabelPageObject("#label");
        label.el().should("contain.text", "Etikett");
        label.description().should("contain.text", "Hjälptext");
        label.formatDescription().should("contain.text", "Formatbeskrivning");
        label.errorMessage().should("contain.text", "Felmeddelande");
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    describe.skip("density", () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <template #default="{ density }">
                        <f-label
                            :id="'label-' + density"
                            :for="'input-' + density"
                        >
                            Etikett
                            <template
                                #description="{ descriptionClass, formatDescriptionClass }"
                            >
                                <span :class="descriptionClass">
                                    Hjälptext
                                </span>
                                <span :class="formatDescriptionClass">
                                    Formatbeskrivning
                                </span>
                            </template>
                            <template #error-message> Felmeddelande </template>
                        </f-label>
                        <input
                            :id="'input-' + density"
                            type="text"
                            class="text-field__input"
                        />
                    </template>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FLabel,
            },
        });

        it(`should be densified`, () => {
            cy.viewport(densityWrapperWidth, densityWrapperHeight);
            cy.mount(DensityComponent);
            cy.toMatchScreenshot();
        });
    });
});
