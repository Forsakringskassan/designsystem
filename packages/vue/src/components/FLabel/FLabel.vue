<script lang="ts">
import { defineComponent, type PropType, provide, useTemplateRef } from "vue";
import { hasSlot } from "../../utils";
import { FIcon } from "../FIcon";
import { tooltipAttachTo } from "../FTooltip";

export default defineComponent({
    name: "FLabel",
    components: {
        FIcon,
    },
    props: {
        /**
         * The id for the form element the label is bound to.
         */
        for: {
            type: String as PropType<string | undefined>,
            required: false,
            default: undefined,
        },
    },
    setup() {
        provide(tooltipAttachTo, useTemplateRef("tooltipAttachTo"));
    },
    data() {
        return {
            descriptionClass: ["label__description"],
            formatDescriptionClass: ["label__description", "label__description--format"],
        };
    },
    computed: {
        forProperty(): string | undefined {
            return this.for;
        },
        hasDefaultSlot(): boolean {
            return hasSlot(this, "default");
        },
        hasErrorMessageSlot(): boolean {
            return hasSlot(this, "error-message");
        },
        hasDescriptionSlot(): boolean {
            return hasSlot(this, "description");
        },
    },
});
</script>

<template>
    <div v-if="$slots.tooltip">
        <div v-if="hasDefaultSlot" ref="tooltipAttachTo">
            <label class="label" :for="forProperty">
                <slot name="default"></slot>
            </label>
        </div>

        <!-- @slot Slot for tooltip. -->
        <slot name="tooltip"></slot>

        <label v-if="hasDescriptionSlot || hasErrorMessageSlot" class="label sr-separator" :for="forProperty">
            <!--
                @slot Optional slot for description.
                @binding {string[]} description-class CSS classes for primary description content.
                @binding {string[]} format-description-class CSS classes for format description.
            -->
            <slot name="description" :description-class :format-description-class></slot>
            <span v-if="hasErrorMessageSlot" class="label__message label__message--error">
                <f-icon class="label__icon--left" name="error"></f-icon>
                <slot name="error-message"></slot>
            </span>
        </label>
    </div>

    <label v-else class="label" :for="forProperty">
        <!-- @slot Slot for label content. -->
        <slot name="default"></slot>
        <!--
            @slot Optional slot for description.
            @binding {string[]} description-class CSS classes for primary description content.
            @binding {string[]} format-description-class CSS classes for format description.
          -->
        <slot name="description" :description-class :format-description-class></slot>
        <span v-if="hasErrorMessageSlot" class="label__message label__message--error">
            <f-icon class="label__icon--left" name="error"></f-icon>
            <!-- @slot Slot for displaying single or several error messages. -->
            <slot name="error-message"></slot>
        </span>
    </label>
</template>
