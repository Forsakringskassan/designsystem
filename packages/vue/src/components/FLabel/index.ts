export { default as FLabel } from "./FLabel.vue";
