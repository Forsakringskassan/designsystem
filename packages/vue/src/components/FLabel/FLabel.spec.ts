import "html-validate/jest";
import { VueWrapper, mount } from "@vue/test-utils";
import FLabel from "./FLabel.vue";

function createWrapper({ slots = {} } = {}): VueWrapper {
    return mount(FLabel, {
        props: { for: "FOR_ID" },
        slots: { ...slots },
        global: {
            stubs: ["f-icon"],
        },
    });
}

describe("snapshots", () => {
    it("should match snapshot with one label element containing the default slot", () => {
        const wrapper = createWrapper({
            slots: {
                default: "LABEL_TEXT",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with one label element containing the default, description and error-message slots", () => {
        const wrapper = createWrapper({
            slots: {
                default: "LABEL_TEXT",
                description: "DESCRIPTION",
                "error-message": "ERRROR_MESSAGE",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with the tooltip slot rendered outside the label element", () => {
        const wrapper = createWrapper({
            slots: {
                default: "LABEL_TEXT",
                tooltip: "TOOLTIP",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with a second label element containing the error-message slot", () => {
        const wrapper = createWrapper({
            slots: {
                default: "LABEL_TEXT",
                tooltip: "TOOLTIP",
                "error-message": "ERROR_MESSAGE",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with a second label element containing the description and error-message slots", () => {
        const wrapper = createWrapper({
            slots: {
                default: "LABEL_TEXT",
                tooltip: "TOOLTIP",
                description: "DESCRIPTION",
                "error-message": "ERROR_MESSAGE",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("html-validate", () => {
    it.each`
        html
        ${"<f-label>Label</f-label>"}
        ${'<f-label><template #tooltip><f-tooltip screen-reader-text="Read more about FLabel"></f-tooltip></template></f-label>'}
        ${"<f-label><template #tooltip><slot></slot></template></f-label>"}
        ${'<f-label for="number-input"></f-label>'}
    `("$html should be valid", ({ html }) => {
        expect.assertions(1);

        expect(html).toHTMLValidate();
    });

    it.each`
        html
        ${"<f-label><template #tooltip><div /></template></f-label>"}
        ${'<f-label for="00-number-input"></f-label>'}
        ${'<f-label for="number input"></f-label>'}
    `("$html should be invalid", ({ html }) => {
        expect.assertions(3);
        let catchedError;

        try {
            expect(html).toHTMLValidate();
        } catch (error) {
            catchedError = error;
        } finally {
            expect(catchedError).toBeDefined();
            expect(catchedError).toMatchSnapshot();
        }
    });
});
