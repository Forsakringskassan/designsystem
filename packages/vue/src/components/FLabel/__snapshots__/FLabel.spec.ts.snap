// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html-validate <f-label for="00-number-input"></f-label> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  Attribute "for" has invalid value "00-number-input" [attribute-allowed-values]]
`;

exports[`html-validate <f-label for="number input"></f-label> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  Attribute "for" has invalid value "number input" [attribute-allowed-values]]
`;

exports[`html-validate <f-label><template #tooltip><div /></template></f-label> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <div> element is not permitted as content under slot "tooltip" (<f-label>) [element-permitted-content]]
`;

exports[`snapshots should match snapshot with a second label element containing the description and error-message slots 1`] = `
<div>
  <div>
    <label
      class="label"
      for="FOR_ID"
    >
      
      LABEL_TEXT
      
    </label>
  </div>
  
  TOOLTIP
  
  <label
    class="label sr-separator"
    for="FOR_ID"
  >
    
    DESCRIPTION
    
    <span
      class="label__message label__message--error"
    >
      <f-icon-stub
        class="label__icon--left"
        library="f"
        name="error"
      />
      
      ERROR_MESSAGE
      
    </span>
  </label>
</div>
`;

exports[`snapshots should match snapshot with a second label element containing the error-message slot 1`] = `
<div>
  <div>
    <label
      class="label"
      for="FOR_ID"
    >
      
      LABEL_TEXT
      
    </label>
  </div>
  
  TOOLTIP
  
  <label
    class="label sr-separator"
    for="FOR_ID"
  >
    
    
    <span
      class="label__message label__message--error"
    >
      <f-icon-stub
        class="label__icon--left"
        library="f"
        name="error"
      />
      
      ERROR_MESSAGE
      
    </span>
  </label>
</div>
`;

exports[`snapshots should match snapshot with one label element containing the default slot 1`] = `
<label
  class="label"
  for="FOR_ID"
>
  
  LABEL_TEXT
  
  
  
  <!--v-if-->
</label>
`;

exports[`snapshots should match snapshot with one label element containing the default, description and error-message slots 1`] = `
<label
  class="label"
  for="FOR_ID"
>
  
  LABEL_TEXT
  
  
  DESCRIPTION
  
  <span
    class="label__message label__message--error"
  >
    <f-icon-stub
      class="label__icon--left"
      library="f"
      name="error"
    />
    
    ERRROR_MESSAGE
    
  </span>
</label>
`;

exports[`snapshots should match snapshot with the tooltip slot rendered outside the label element 1`] = `
<div>
  <div>
    <label
      class="label"
      for="FOR_ID"
    >
      
      LABEL_TEXT
      
    </label>
  </div>
  
  TOOLTIP
  
  <!--v-if-->
</div>
`;
